scheduler/core/components/collector/__init__.py

```diff
@@ -526,37 +570,122 @@ def time_accounting(self,
             # We set to None is the whole night is to be done.
             end_timeslot_bound = end_timeslot_bounds.get(plan.site) if end_timeslot_bounds is not None else None
 
-            # Ensure visits are processed in order of starting time slot.
-            for visit in sorted(plan.visits, key=lambda v: v.start_time_slot):
-                # If this visit's starting time slot is at least the end timeslot bound, then we stop
-                # processing this and all further visits.
-                if end_timeslot_bound is not None and visit.start_time_slot >= end_timeslot_bound:
-                    break
-
-                # We process this visit.
-                # Update Observation from Collector.
-                observation = self.get_observation(visit.obs_id)
-                obs_seq = observation.sequence
-
-                # Check if the Observation has been completely observed.
-                if visit.atom_end_idx == len(obs_seq) - 1:
-                    logger.info(f'Marking observation complete: {observation.id.id}')
-                    observation.status = ObservationStatus.OBSERVED
+            # print(f'\ntime_accounting')
+            # print(f'{plan.site} {end_timeslot_bound}')
+
+            group_visits = []
+            for ii, visit in enumerate(sorted(plan.visits, key=lambda v: v.start_time_slot)):
+                obs = self.get_observation(visit.obs_id)
+                group = self._get_group(obs)
+                # print(
+                #     f'{ii} {visit.obs_id.id} {group.unique_id.id} {visit.start_time_slot} {visit.start_time_slot + visit.time_slots - 1}')
+                if ii > 1 and group.is_scheduling_group() and group == group_visits[-1].group:
```

Comment:
I think I see what is happening here, but the logic is a bit dangerous: `group_visits` is declared empty and indexing it here could cause an `IndexError`. I think the `ii > 1` circumvents this, but we should look at something more robust.